ext/drx/drx.c

```diff
@@ -1372,3 +1405,1117 @@ drx_tail_pad_block(void *drcontext, instrlist_t *ilist)
+ * mov $0x00000002 -> %ecx
+ * kmovw %ecx -> %k0
+ * kandnw %k0 %k1 -> %k1
+ * skip1:
+ *
+ * [..]
+ *
+ * Do the same as above for the last element:
+ *
+ * kmovw %k1 -> %ecx
+ * test %ecx $0x00008000
+ * jz <skip15>
+ * vextracti32x4 {%k0} $0x03 %zmm1 -> %xmm2
+ * vpextrd %xmm2 $0x03 -> %ecx
+ * mov (%rax,%rcx,4)[4byte] -> %ecx
+ * vextracti32x4 {%k0} $0x03 %zmm0 -> %xmm2
+ * vpinsrd %xmm2 %ecx $0x03 -> %xmm2
+ * vinserti32x4 {%k0} $0x03 %zmm0 %xmm2 -> %zmm0
+ * mov $0x00008000 -> %ecx
+ * kmovw %ecx -> %k0
+ * kandnw %k0 %k1 -> %k1
+ * skip15:
+ *
+ * Finally, clear the entire mask register, even
+ * the parts that are not used as a mask:
+ *
+ * kxorq %k1 %k1 -> %k1
+ *
+ * AVX-512 vpscatterdd, vscatterdps, vpscatterdq, vscatterdpd,
+ * vpscatterqd, vscatterqps, vpscatterqq, vscatterqpd:
+ *
+ * vpscatterdd {%k1} %zmm0 -> (%rcx,%zmm1,4)[4byte] %k1 sequence laid out here, others
+ * are similar:
+ *
+ * Extract mask bit:
+ *
+ * kmovw %k1 -> %edx
+ * Test mask bit:
+ * test %edx $0x00000001
+ *
+ * Skip element if mask not set:
+ *
+ * jz <skip0>
+ *
+ * Extract index dword. qword versions use vpextrq:
+ *
+ * vextracti32x4 {%k0} $0x00 %zmm1 -> %xmm2
+ * vpextrd %xmm2 $0x00 -> %edx
+ *
+ * Extract scalar value dword. qword versions use vpextrq:
+ *
+ * vextracti32x4 {%k0} $0x00 %zmm0 -> %xmm2
+ * vpextrd %xmm2 $0x00 -> %ebx
+ *
+ * Store scalar value:
+ *
+ * mov %ebx -> (%rcx,%rdx,4)[4byte]
+ *
+ * Set mask bit to zero:
+ *
+ * mov $0x00000001 -> %edx
+ * kmovw %edx -> %k0
+ * kandnw %k0 %k1 -> %k1
+ * skip0:
+ *
+ * Do the same as above for the next element:
+ *
+ * kmovw %k1 -> %edx
+ * test %edx $0x00000002
+ * jz <skip1>
+ * vextracti32x4 {%k0} $0x00 %zmm1 -> %xmm2
+ * vpextrd %xmm2 $0x01 -> %edx
+ * vextracti32x4 {%k0} $0x00 %zmm0 -> %xmm2
+ * vpextrd %xmm2 $0x01 -> %ebx
+ * mov %ebx -> (%rcx,%rdx,4)[4byte]
+ * mov $0x00000002 -> %edx
+ * kmovw %edx -> %k0
+ * kandnw %k0 %k1 -> %k1
+ * skip1:
+ *
+ * [..]
+ *
+ *
+ * Do the same as above for the last element:
+ *
+ * kmovw %k1 -> %edx
+ * test %edx $0x00008000
+ * jz <skip15>
+ * vextracti32x4 {%k0} $0x03 %zmm1 -> %xmm2
+ * vpextrd %xmm2 $0x03 -> %edx
+ * vextracti32x4 {%k0} $0x03 %zmm0 -> %xmm2
+ * vpextrd %xmm2 $0x03 -> %ebx
+ * mov %ebx -> (%rcx,%rdx,4)[4byte]
+ * mov $0x00008000 -> %edx
+ * kmovw %edx -> %k0
+ * kandnw %k0 %k1 -> %k1
+ * skip15:
+ *
+ * Finally, clear the entire mask register, even
+ * the parts that are not used as a mask:
+ *
+ * kxorq %k1 %k1 -> %k1
+ */
+bool
+drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
+{
+#ifdef X86
+    instr_t *instr, *next_instr, *first_app = NULL;
+    bool delete_rest = false;
+#endif
+    if (expanded != NULL)
+        *expanded = false;
+    if (drmgr_current_bb_phase(drcontext) != DRMGR_PHASE_APP2APP) {
+        USAGE_ERROR("drutil_expand_rep_string* must be called from "
+                    "drmgr's app2app phase");
+        return false;
+    }
+#ifdef X86
```

Comment:
Wouldn't putting into a separate block be cross-platform?

Reply:
Not sure what you mean here. Can you please re-phrase?

Reply:
The code here separating a scatter instruction into its own basic block has no need to be x86-only: nothing it's doing is x86-specific.

Reply:
O I see, I could pull this into the cross-platform part, but since AArch64 is completely unsupported, I think I am adding a small comment here mentioning that this should be shared in the future.